unified-scheduler-logic/src/lib.rs

```diff
@@ -562,91 +670,305 @@ impl Default for UsageQueueInner {
+            self.pop_usage_from_task()
         } else {
             None
         }
     }
 
     fn push_blocked_usage_from_task(&mut self, usage_from_task: UsageFromTask) {
-        assert_matches!(self.current_usage, Some(_));
-        self.blocked_usages_from_tasks.push_back(usage_from_task);
+        assert_matches!(self.current_usage(), Some(_));
+        self.push_usage_from_task(usage_from_task);
     }
 
     #[must_use]
-    fn pop_unblocked_readonly_usage_from_task(&mut self) -> Option<UsageFromTask> {
+    fn pop_lockable_readonly_usage_from_task(&mut self) -> Option<UsageFromTask> {
         if matches!(
-            self.blocked_usages_from_tasks.front(),
+            self.peek_usage_from_task(),
             Some((RequestedUsage::Readonly, _))
         ) {
-            assert_matches!(self.current_usage, Some(Usage::Readonly(_)));
-            self.blocked_usages_from_tasks.pop_front()
+            assert_matches!(self.current_usage(), Some(RequestedUsage::Readonly));
+            self.pop_usage_from_task()
         } else {
             None
         }
     }
 
-    fn has_no_blocked_usage(&self) -> bool {
-        self.blocked_usages_from_tasks.is_empty()
+    fn current_usage(&self) -> Option<RequestedUsage> {
+        match self {
+            Self::Fifo { current_usage, .. } => {
+                current_usage.as_ref().map(|usage| usage.requested_usage())
+            }
+            Self::Priority { current_usage, .. } => {
+                current_usage.as_ref().map(|usage| usage.requested_usage())
+            }
+        }
+    }
+
+    #[cfg(test)]
+    fn update_current_usage(&mut self, requested_usage: RequestedUsage, task: &Task) {
+        match self {
+            Self::Fifo { current_usage, .. } => {
+                *current_usage = Some(FifoUsage::from(requested_usage));
+            }
+            Self::Priority { current_usage, .. } => {
+                *current_usage = Some(PriorityUsage::from(task.clone(), requested_usage));
+            }
+        }
+    }
+
+    fn pop_usage_from_task(&mut self) -> Option<UsageFromTask> {
+        match self {
+            Self::Fifo {
+                blocked_usages_from_tasks,
+                ..
+            } => blocked_usages_from_tasks.pop_front(),
+            Self::Priority {
+                blocked_usages_from_tasks,
+                ..
+            } => blocked_usages_from_tasks.pop_first_usage_from_task(),
+        }
+    }
+
+    fn push_usage_from_task(&mut self, usage_from_task: UsageFromTask) {
+        match self {
+            Self::Fifo {
+                blocked_usages_from_tasks,
+                ..
+            } => blocked_usages_from_tasks.push_back(usage_from_task),
+            Self::Priority {
+                blocked_usages_from_tasks,
+                ..
+            } => blocked_usages_from_tasks.insert_usage_from_task(usage_from_task),
+        }
+    }
+
+    fn peek_usage_from_task(&self) -> Option<&UsageFromTask> {
```

Comment:
`peek_blocked`? this doesn't peek at the current usage, it's just the blocked

Reply:
done: ea8cf75ad5e6b78c50ee802035929878b20db63a